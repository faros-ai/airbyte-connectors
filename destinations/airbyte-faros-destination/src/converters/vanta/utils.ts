import fs from 'fs-extra';
import path from 'path';

// Function to find a specific parent directory
function findParentDirectory(
  currentDir: string,
  targetDirName: string
): string | null {
  let currentPath = currentDir;
  while (currentPath !== '/') {
    if (path.basename(currentPath) === targetDirName) {
      return currentPath;
    }
    currentPath = path.dirname(currentPath);
  }
  return null;
}

export function getQueryFromName(name: string): string {
  const crt_dir = __dirname;
  // We go back to the part of dir which ends with name:
  const targetDirPath = findParentDirectory(
    crt_dir,
    'airbyte-faros-destination'
  );
  if (!targetDirPath) {
    throw new Error('airbyte-faros-destination directory not found');
  }

  const gql_dir = path.join(targetDirPath, 'resources');

  const fn = `${name}.gql`;
  // Check if fn is in gql_dir:
  if (!fs.existsSync(`${gql_dir}/${fn}`)) {
    throw new Error(`Query ${fn} not found in ${gql_dir}`);
  }
  // We access the directory resources/gql_queries to get the query
  return fs.readFileSync(`${gql_dir}/${fn}`, 'utf8');
}

export function looksLikeGitCommitSha(sha: string): boolean {
  return /^[a-f0-9]{40}$/i.test(sha);
}
