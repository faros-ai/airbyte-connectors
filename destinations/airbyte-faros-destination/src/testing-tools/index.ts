export * from './cli';
export * from './testing-tools';
export * from './utils';
