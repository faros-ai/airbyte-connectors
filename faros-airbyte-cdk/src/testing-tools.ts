import fs from 'fs';
import {Dictionary} from 'ts-essentials';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
} from '.';

export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}
