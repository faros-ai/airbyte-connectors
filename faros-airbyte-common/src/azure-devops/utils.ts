import {User} from './types';

/**
 * Gets a unique identifier for an Azure DevOps user.
 * Tries principalName first, then falls back to uniqueName.
 * This handles differences between Azure DevOps Cloud and Server.
 *
 * @param user - The Azure DevOps user object
 * @returns The user's unique identifier, or undefined if neither field exists
 */
export function getUserIdentifier(user: User): string | undefined {
  if ('principalName' in user && user.principalName) {
    return user.principalName;
  }
  if ('uniqueName' in user && user.uniqueName) {
    return user.uniqueName;
  }
  return undefined;
}
