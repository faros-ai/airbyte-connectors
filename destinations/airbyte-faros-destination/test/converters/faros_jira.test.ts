import {getLocal} from 'mockttp';

import {initMockttp, tempConfig} from '../testing-tools';
import {destinationWriteTest} from './utils';

describe('faros_jira', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let configPath: string;

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig({
      api_url: mockttp.url,
      log_records: true,
    });
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    await destinationWriteTest({
      configPath,
      catalogPath: 'test/resources/faros_jira/catalog.json',
      inputRecordsPath: 'faros_jira/all-streams.log',
      checkRecordsData: (records) => expect(records).toMatchSnapshot(),
    });
  });

  test('faros_issues', async () => {
    await destinationWriteTest({
      configPath,
      catalogPath: 'test/resources/faros_jira/catalog.json',
      inputRecordsPath: 'faros_jira/faros_issues.log',
      checkRecordsData: (records) => expect(records).toMatchSnapshot(),
    });
  });
});
