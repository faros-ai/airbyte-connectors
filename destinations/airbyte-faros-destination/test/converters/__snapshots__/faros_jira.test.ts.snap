// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`faros_jira faros_issues 1`] = `
Array [
  "Processed 1 records",
  "Processed records by stream: {\\\\\\"mytestsource__jira__faros_issues\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 20 records",
  "Would write records by model: {\\\\\\"tms_Epic\\\\\\":1,\\\\\\"tms_Label\\\\\\":2,\\\\\\"tms_SprintHistory\\\\\\":11,\\\\\\"tms_Task\\\\\\":1,\\\\\\"tms_TaskAssignment\\\\\\":1,\\\\\\"tms_TaskDependency\\\\\\":1,\\\\\\"tms_TaskProjectRelationship\\\\\\":1,\\\\\\"tms_TaskTag\\\\\\":2}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`faros_jira faros_issues 2`] = `
Array [
  Object {
    "tms_Task": Object {
      "additionalFields": Array [
        Object {
          "name": "Components",
          "value": "[]",
        },
      ],
      "createdAt": "2021-10-07T20:00:08.396Z",
      "creator": Object {
        "source": "Jira",
        "uid": "6025c1954fb89e0070d00ae3",
      },
      "description": "PLAYG Epic 1 Description",
      "epic": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
      "name": "PLAYG Epic 1 Summary",
      "parent": null,
      "points": null,
      "priority": "Medium",
      "resolutionStatus": null,
      "resolvedAt": null,
      "source": "Jira",
      "sourceSystemId": "10596",
      "sprint": Object {
        "source": "Jira",
        "uid": "55",
      },
      "status": Object {
        "category": "Todo",
        "detail": "To Do",
      },
      "statusChangedAt": "2021-10-07T20:00:08.396Z",
      "statusChangelog": Array [
        Object {
          "changedAt": "2021-10-07T20:00:08.396Z",
          "status": Object {
            "category": "Todo",
            "detail": "To Do",
          },
        },
      ],
      "type": Object {
        "category": "Custom",
        "detail": "Epic",
      },
      "uid": "PLAYG-6",
      "updatedAt": "2024-01-24T14:56:25.365Z",
      "url": "https://faros-ai.atlassian.net/browse/PLAYG-6",
    },
  },
  Object {
    "tms_TaskProjectRelationship": Object {
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_Epic": Object {
      "createdAt": "2021-10-07T20:00:08.396Z",
      "description": "PLAYG Epic 1 Description",
      "name": "PLAYG Epic 1 Summary",
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "source": "Jira",
      "status": Object {
        "category": "Todo",
        "detail": "To Do",
      },
      "uid": "PLAYG-6",
      "updatedAt": "2024-01-24T14:56:25.365Z",
    },
  },
  Object {
    "tms_TaskAssignment": Object {
      "assignedAt": "2021-10-07T18:04:10.236Z",
      "assignee": Object {
        "source": "Jira",
        "uid": "6025c1954fb89e0070d00ae3",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_TaskDependency": Object {
      "blocking": true,
      "dependencyType": Object {
        "category": "BlockedBy",
        "detail": "is blocked by",
      },
      "dependentTask": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
      "fulfillingTask": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
      "fulfillingType": Object {
        "category": "Blocks",
        "detail": "blocks",
      },
      "source": "Jira",
    },
  },
  Object {
    "tms_Label": Object {
      "name": "l0",
      "source": "Jira",
    },
  },
  Object {
    "tms_TaskTag": Object {
      "label": Object {
        "name": "l0",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_Label": Object {
      "name": "l1",
      "source": "Jira",
    },
  },
  Object {
    "tms_TaskTag": Object {
      "label": Object {
        "name": "l1",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-16T22:42:42.308Z",
      "removedAt": "2021-09-16T22:45:30.905Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "15",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-16T22:47:11.885Z",
      "removedAt": "2021-09-16T22:47:23.663Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "16",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T06:36:26.329Z",
      "removedAt": "2021-09-17T06:37:15.206Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "17",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T07:12:32.830Z",
      "removedAt": "2021-09-22T16:57:44.008Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "18",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-22T16:57:44.008Z",
      "removedAt": "2021-09-22T20:38:03.327Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "19",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T07:10:10.556Z",
      "removedAt": "2021-09-17T07:12:32.830Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "20",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-22T20:38:03.327Z",
      "removedAt": "2021-10-01T22:16:08.307Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "21",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-10-01T22:16:08.673Z",
      "removedAt": "2022-06-13T22:49:44.016Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "23",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2022-06-16T15:34:47.608Z",
      "removedAt": "2022-06-16T20:50:28.697Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "27",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2024-03-08T17:59:28.180Z",
      "removedAt": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "55",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2024-03-08T17:30:59.921Z",
      "removedAt": "2024-03-08T17:59:28.180Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "66",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-6",
      },
    },
  },
]
`;

exports[`faros_jira process records from all streams 1`] = `
Array [
  "Processed 17 records",
  "Processed records by stream: {\\\\\\"mytestsource__jira__faros_board_issues\\\\\\":1,\\\\\\"mytestsource__jira__faros_boards\\\\\\":2,\\\\\\"mytestsource__jira__faros_issue_additional_fields\\\\\\":1,\\\\\\"mytestsource__jira__faros_issue_pull_requests\\\\\\":1,\\\\\\"mytestsource__jira__faros_issues\\\\\\":1,\\\\\\"mytestsource__jira__faros_project_version_issues\\\\\\":1,\\\\\\"mytestsource__jira__faros_project_versions\\\\\\":1,\\\\\\"mytestsource__jira__faros_projects\\\\\\":1,\\\\\\"mytestsource__jira__faros_sprint_reports\\\\\\":2,\\\\\\"mytestsource__jira__faros_sprints\\\\\\":2,\\\\\\"mytestsource__jira__faros_team_memberships\\\\\\":2,\\\\\\"mytestsource__jira__faros_teams\\\\\\":1,\\\\\\"mytestsource__jira__faros_users\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 47 records",
  "Would write records by model: {\\\\\\"tms_Project\\\\\\":1,\\\\\\"tms_ProjectReleaseRelationship\\\\\\":1,\\\\\\"tms_Release\\\\\\":1,\\\\\\"tms_Sprint\\\\\\":2,\\\\\\"tms_SprintBoardRelationship\\\\\\":2,\\\\\\"tms_SprintHistory\\\\\\":25,\\\\\\"tms_Task\\\\\\":1,\\\\\\"tms_TaskAssignment\\\\\\":1,\\\\\\"tms_TaskBoard\\\\\\":2,\\\\\\"tms_TaskBoardProjectRelationship\\\\\\":2,\\\\\\"tms_TaskBoardRelationship\\\\\\":1,\\\\\\"tms_TaskProjectRelationship\\\\\\":1,\\\\\\"tms_TaskPullRequestAssociation\\\\\\":1,\\\\\\"tms_TaskReleaseRelationship\\\\\\":1,\\\\\\"tms_Task__Update\\\\\\":1,\\\\\\"tms_Team\\\\\\":1,\\\\\\"tms_TeamMembership\\\\\\":2,\\\\\\"tms_User\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`faros_jira process records from all streams 2`] = `
Array [
  Object {
    "tms_TaskBoardRelationship": Object {
      "board": Object {
        "source": "Jira",
        "uid": "48",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-185",
      },
    },
  },
  Object {
    "tms_TaskBoard": Object {
      "name": "Tasks without a board in project PLAYG",
      "source": "Jira",
      "uid": "faros-tasks-with-no-board-PLAYG",
    },
  },
  Object {
    "tms_TaskBoardProjectRelationship": Object {
      "board": Object {
        "source": "Jira",
        "uid": "faros-tasks-with-no-board-PLAYG",
      },
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "source": "Jira",
    },
  },
  Object {
    "tms_TaskBoard": Object {
      "name": "PLAYG board",
      "source": "Jira",
      "uid": "21",
    },
  },
  Object {
    "tms_TaskBoardProjectRelationship": Object {
      "board": Object {
        "source": "Jira",
        "uid": "21",
      },
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "source": "Jira",
    },
  },
  Object {
    "tms_TaskPullRequestAssociation": Object {
      "pullRequest": Object {
        "number": 2069,
        "repository": Object {
          "name": "airbyte-connectors",
          "organization": Object {
            "source": "GitHub",
            "uid": "faros-ai",
          },
        },
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-153",
      },
    },
  },
  Object {
    "tms_Task": Object {
      "additionalFields": Array [
        Object {
          "name": "Components",
          "value": "[]",
        },
      ],
      "createdAt": "2021-09-16T22:42:13.869Z",
      "creator": Object {
        "source": "Jira",
        "uid": "60d6b0c0c90cb2006892d7fb",
      },
      "description": null,
      "epic": null,
      "name": "task 1",
      "parent": null,
      "points": 1,
      "priority": "Medium",
      "resolutionStatus": null,
      "resolvedAt": null,
      "source": "Jira",
      "sourceSystemId": "10456",
      "sprint": Object {
        "source": "Jira",
        "uid": "55",
      },
      "status": Object {
        "category": "Todo",
        "detail": "To Do",
      },
      "statusChangedAt": "2024-03-08T17:52:41.851Z",
      "statusChangelog": Array [
        Object {
          "changedAt": "2021-09-16T22:42:13.869Z",
          "status": Object {
            "category": "Todo",
            "detail": "To Do",
          },
        },
        Object {
          "changedAt": "2021-09-16T22:43:43.077Z",
          "status": Object {
            "category": "InProgress",
            "detail": "In Progress",
          },
        },
        Object {
          "changedAt": "2022-03-15T07:19:03.653Z",
          "status": Object {
            "category": "InProgress",
            "detail": "QA",
          },
        },
        Object {
          "changedAt": "2024-03-08T17:52:41.851Z",
          "status": Object {
            "category": "Todo",
            "detail": "To Do",
          },
        },
      ],
      "type": Object {
        "category": "Story",
        "detail": "Story",
      },
      "uid": "PLAYG-1",
      "updatedAt": "2024-03-08T17:59:28.167Z",
      "url": "https://faros-ai.atlassian.net/browse/PLAYG-1",
    },
  },
  Object {
    "tms_TaskProjectRelationship": Object {
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_TaskAssignment": Object {
      "assignedAt": "2021-10-07T18:04:10.236Z",
      "assignee": Object {
        "source": "Jira",
        "uid": "6025c1954fb89e0070d00ae3",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-16T22:42:42.308Z",
      "removedAt": "2021-09-16T22:45:30.905Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "15",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-16T22:47:11.885Z",
      "removedAt": "2021-09-16T22:47:23.663Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "16",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T06:36:26.329Z",
      "removedAt": "2021-09-17T06:37:15.206Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "17",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T07:12:32.830Z",
      "removedAt": "2021-09-22T16:57:44.008Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "18",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-22T16:57:44.008Z",
      "removedAt": "2021-09-22T20:38:03.327Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "19",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-17T07:10:10.556Z",
      "removedAt": "2021-09-17T07:12:32.830Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "20",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-09-22T20:38:03.327Z",
      "removedAt": "2021-10-01T22:16:08.307Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "21",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2021-10-01T22:16:08.673Z",
      "removedAt": "2022-06-13T22:49:44.016Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "23",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2022-06-16T15:34:47.608Z",
      "removedAt": "2022-06-16T20:50:28.697Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "27",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2024-03-08T17:59:28.180Z",
      "removedAt": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "55",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedAt": "2024-03-08T17:30:59.921Z",
      "removedAt": "2024-03-08T17:59:28.180Z",
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "66",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "Completed",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-44",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "Completed",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-45",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "CompletedOutsideSprint",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-3",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "CompletedOutsideSprint",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-26",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-9",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-15",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "39",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-22",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-4",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-20",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": 9,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-24",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-29",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-38",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": true,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "NotCompleted",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-40",
      },
    },
  },
  Object {
    "tms_SprintHistory": Object {
      "addedDuringSprint": null,
      "points": null,
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "51",
      },
      "status": Object {
        "category": "Removed",
      },
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-11",
      },
    },
  },
  Object {
    "tms_Sprint": Object {
      "closedAt": "2024-05-13T07:10:11.135Z",
      "endedAt": "2024-05-19T22:39:39.000Z",
      "name": "Another Sprint",
      "openedAt": null,
      "source": "Jira",
      "startedAt": "2024-05-13T07:09:41.502Z",
      "state": "Closed",
      "uid": "75",
    },
  },
  Object {
    "tms_SprintBoardRelationship": Object {
      "board": Object {
        "source": "Jira",
        "uid": "49",
      },
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "75",
      },
    },
  },
  Object {
    "tms_Sprint": Object {
      "closedAt": null,
      "endedAt": null,
      "name": "PLAYG Sprint 20",
      "openedAt": null,
      "source": "Jira",
      "startedAt": null,
      "state": "Future",
      "uid": "55",
    },
  },
  Object {
    "tms_SprintBoardRelationship": Object {
      "board": Object {
        "source": "Jira",
        "uid": "49",
      },
      "source": "Jira",
      "sprint": Object {
        "source": "Jira",
        "uid": "55",
      },
    },
  },
  Object {
    "tms_Project": Object {
      "description": "Playground project description",
      "name": "Playground",
      "source": "Jira",
      "sourceSystemId": "10011",
      "uid": "PLAYG",
    },
  },
  Object {
    "tms_User": Object {
      "emailAddress": "john.doe@example.com",
      "inactive": false,
      "name": "John Doe",
      "source": "Jira",
      "uid": "61f99bc0acd4cb0069cfd675",
    },
  },
  Object {
    "tms_TaskReleaseRelationship": Object {
      "release": Object {
        "source": "Jira",
        "uid": "10006",
      },
      "source": "Jira",
      "task": Object {
        "source": "Jira",
        "uid": "PLAYG-4",
      },
    },
  },
  Object {
    "tms_Release": Object {
      "description": null,
      "name": "Test Release",
      "releasedAt": "2021-11-19T00:00:00.000Z",
      "source": "Jira",
      "startedAt": "2021-11-15T00:00:00.000Z",
      "uid": "10006",
    },
  },
  Object {
    "tms_ProjectReleaseRelationship": Object {
      "project": Object {
        "source": "Jira",
        "uid": "PLAYG",
      },
      "release": Object {
        "source": "Jira",
        "uid": "10006",
      },
      "source": "Jira",
    },
  },
  Object {
    "tms_Team": Object {
      "name": "Test",
      "parentTeam": Object {
        "uid": "all_teams",
      },
      "source": "Jira",
      "teamChain": Array [
        "123ad121-1122-45c5-8d0b-0681b9233ea9",
        "all_teams",
      ],
      "uid": "123ad121-1122-45c5-8d0b-0681b9233ea9",
    },
  },
  Object {
    "tms_TeamMembership": Object {
      "member": Object {
        "source": "Jira",
        "uid": "12334c037692da100700011",
      },
      "source": "Jira",
      "team": Object {
        "source": "Jira",
        "uid": "123ad121-1122-45c5-8d0b-0681b9233ea9",
      },
    },
  },
  Object {
    "tms_TeamMembership": Object {
      "member": Object {
        "source": "Jira",
        "uid": "3214dfa413fc70069d12323",
      },
      "source": "Jira",
      "team": Object {
        "source": "Jira",
        "uid": "123ad121-1122-45c5-8d0b-0681b9233ea9",
      },
    },
  },
  Object {
    "tms_Task__Update": Object {
      "mask": Array [
        "additionalFields",
      ],
      "patch": Object {
        "additionalFields": Array [
          Object {
            "name": "Custom String Field",
            "value": "Custom",
          },
          Object {
            "name": "Custom Number Field",
            "value": "1",
          },
        ],
      },
      "source": "Jira",
      "where": Object {
        "source": "Jira",
        "uid": "PLAYG-1",
      },
    },
  },
]
`;
